File: problem-002/problem-2.kt
fun generateFibonacci(x: Int, y: Int): Sequence<Int> = sequence {
    yield(x)
    yieldAll(generateFibonacci(y, x + y))
}

fun main() = println(

    generateFibonacci(1, 2)
        .takeWhile { it <= 4_000_000 }
        .filter { it % 2 == 0 }
        .sum()

)

File: problem-002/problem-2.rs
use std::iter;

fn main() {
	let mut fib1: i32 = 1;
	let mut fib2: i32 = 2;
	let fibs = iter::repeat_with(|| {
		let hold = fib1;
		fib1 = fib2;
		fib2 += hold;

		hold
	});

	println!("{}",
		fibs
			.take_while(|&n| n < 4_000_000)
			.filter(|&n| n % 2 == 0)
			.sum::<i32>()
	);
}